### DISCUSSION TITLE
Creating all components from the command line

### DISCUSSION THREAD

**Discussion by xperseguers** [WeblateOrg/weblate#7515] [Q&A] [answered]
Hi there!

After manually creating a few projects in Weblate (and starting to move away from Pootle), I try to create a script to kickstart a TYPO3 extension project (XLIFF-based) for localizing it.

The structure of a TYPO3 extension is:

```
|-- ...
|-- Resources
:   |-- Private
:   :   |-- Language
:   :   :   |-- file1.xlf
:   :   :   |-- file2.xlf
:   :   :   :
```

These are English source files and expected translations are either expected to be stored as e.g., `de.file1.xlf`, `fr.file1.xlf` or packaged into some TYPO3-specific zip file and made available for download.

I want to go that way, and do not want Weblate to interfere with my extension repository. So my idea is to take care of copying English source files to the correct "weblate" data directory like I do already with other projects where I need to support custom localization format.

1. I manually clone my repository to `/home/weblate/git-clones/PROJECT.git`
2. I create a Weblate project using the API:
   ```
   curl \
   	--data-binary "{\"name\":\"$PROJECT\",\"slug\":\"$PROJECT\",\"web\":\"$WEB\"}" \
   	-H "Content-Type: application/json" \
   	-H "Authorization: Token $TOKEN" \
   	https://weblate.example.com/api/projects/
   ```
3. Then I import all (English) files as upcoming components:

```
WEBLATE_REPO=/home/weblate/data/vcs/PROJECT
pushd $WEBLATE_REPO >/dev/null
git init --initial-branch=main
popd >/dev/null

pushd $GIT_REPO/Resources/Private/Language/ >/dev/null
FILES=$(find . -name \*.xlf | cut -b3-)
for F in $FILES; do
	COMPONENT=$(echo $F | sed 's/\.xlf//')
	mkdir $WEBLATE_REPO/$COMPONENT
	cp $F $WEBLATE_REPO/$COMPONENT/en.xlf
done
popd >/dev/null

pushd $WEBLATE_REPO >/dev/null
git add .
git commit \
	-m "[TASK] Initialize project" \
	-m "Git repository: $GIT_REPO"
popd >/dev/null
```

So far so good! I now have this in `/home/weblate/data/vcs/PROJECT`:

```
.
├── locallang_db
│   └── en.xlf
└── locallang_modcalendar
    └── en.xlf
```

I now need to automatically create the components "locallang_db" and "locallang_modcalendar", which I try to do from the command line using `weblate` as described in [the documentation](https://docs.weblate.org/en/latest/admin/management.html#import-project):

```bash
$ weblate import_project \
    PROJECT weblate://PROJECT main \
    '**/*.xlf'

WARNING Handled exception: ValueError: not enough values to unpack (expected 2, got 1)
Traceback (most recent call last):
  File "/home/weblate/weblate-env/bin/weblate", line 8, in <module>
    sys.exit(main())
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/weblate/runner.py", line 34, in main
    utility.execute()
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/django/core/management/__init__.py", line 419, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/django/core/management/base.py", line 373, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/weblate/utils/management/base.py", line 42, in execute
    return super().execute(*args, **options)
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/django/core/management/base.py", line 417, in execute
    output = self.handle(*args, **options)
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/weblate/trans/management/commands/import_project.py", line 245, in handle
    component = Component.objects.get_linked(repo)
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/weblate/trans/models/component.py", line 201, in get_linked
    project, component = val[10:].split("/", 1)
ValueError: not enough values to unpack (expected 2, got 1)
```

I don't get how to do it. For other projects, I manually created a first component without any git repository and then created other components based specifying the `weblate:/` URI of the main component.

I'm fine doing it using some API if it works better, I'm fine as well taking care of creating components myself, without any automatic discovery and specifying one my files as "main component" but I really don't want to do all the configuration task manually in the Web UI as it is way too much work.

Thanks for hints.

**Comment by xperseguers** [comment, marked as answer]
OK, found a way to import the main "component". I arbitrarily set it to be the file `locallang.xlf` in my example as I know it always exists (well, at least in my projects and 99% of all TYPO3 extensions :D):

```bash
pushd $GIT_REPO/Resources/Private/Language/ >/dev/null
if [ ! -f locallang.xlf ]; then
	echo "Could not find locallang.xlf in $(pwd)"
	exit 3
fi

# Create the project
JSON=$( jq -n \
	--arg name "$NAME" \
	--arg slug "$PROJECT" \
	--arg web "$WEB" \
	'$ARGS.named' )
curl \
	--data-binary "$JSON" \
	-H "Content-Type: application/json" \
	-H "Authorization: Token $TOKEN" \
	https://weblate.example.com/api/projects/

# Create main component for locallang.xlf
rm -rf /tmp/$PROJECT
mkdir /tmp/$PROJECT
cp locallang.xlf /tmp/$PROJECT/en.xlf

pushd /tmp/$PROJECT >/dev/null
curl \
	--form docfile=@en.xlf \
	--form name=locallang \
	--form slug=locallang \
	--form file_format=xliff \
	--form filemask=*.xlf \
	--form new_base=en.xlf \
	--form new_lang=add \
	-H "Authorization: Token $TOKEN" \
	https://weblate.example.com/api/projects/$PROJECT/components/
popd >/dev/null

rm -rf /tmp/$PROJECT
```

This will effectively create a git repository component under `/home/weblate/data/vcs/PROJECT/locallang/` with following file structure:

```
.
├── locallang
│   └── en.xlf
└── README.md
```

From there, I should be able to copy other XLIFF files in that repository and add the new components.

And indeed, this works, so the end of my script reads:

```bash
# Copy other files
FILES=$(find . -name \*.xlf | cut -b3-)
for F in $FILES; do
	COMPONENT=$(echo $F | sed 's/\.xlf//')
	if [ "$COMPONENT" != "locallang" ]; then
		mkdir $WEBLATE_REPO/$COMPONENT
		cp $F $WEBLATE_REPO/$COMPONENT/en.xlf
	fi	
done
popd >/dev/null

# WEBLATE_REPO is /home/weblate/data/vcs/PROJECT/locallang in this example
pushd $WEBLATE_REPO >/dev/null
git add .
git commit \
	-m "[TASK] Initialize project" \
	-m "Git repository: $GIT_REPO"
popd >/dev/null


# Automatically discover other files
weblate import_project \
	$PROJECT weblate://$PROJECT main \
	'**/*.xlf'
```

which results in the end with:

```
INFO Found 3 matching files
INFO Found 3 components
INFO Found 1 languages
INFO PROJECT/locallang: Creating component locallang_db
INFO PROJECT/locallang: Creating component locallang_modcalendar
```

**Comment by nijel** [comment]
It might be better to use [component discovery add-on](https://docs.weblate.org/en/latest/admin/addons.html#component-discovery) instead of creating components on the command line.

**Reply by xperseguers** [reply]
@nijel Well, this is exactly what I’ve done, haven’t I? Only that as reported in the first post, it doesn’t seem to work until a main component is present. In the very last part of my solution I do use the component discovery as seen in the output.
